### PR TITLE
New robust mesh balancing algorithm ("new-adapt")

### PR DESCRIPTION
**Pull request by jobordner**
This branch implements a new and improved mesh balancing algorithm ("mesh balancing" here refers to maintaining a mesh hierarchy without "level-jumps", and not to dynamic load balancing.) The algorithm was originally developed by Phil Miller in his 2016 Ph.D. thesis (http://hdl.handle.net/2142/95305), and this is the first time it has been implemented.

The motivation for implementing this approach was that the previous balancing algorithm, which used quiescence detection (QD), proved unreliable and resistant to debugging, and could lead to program failures with fatal "level-jump" errors in refresh operations, especially in larger runs. This new algorithm, which uses a global reduction instead of QD, has yet to fail in tests.

### CODE REVIEW

**Comment by mabruzzo** [comment]
If you have a chance in the next few days, would you mind resetting the diff with the main branch now that PR #97 has been merged? 

[This](https://gist.github.com/brandur/10914112) explains a quick procedure you can follow to do that (it essentially involves pushing a blank commit and then removing it).

**Comment by peeples** [comment]
@bwoshea @mabruzzo @fearmayo @jwise77 While we still only need two approvals to merge, I've assigned all four of you to this PR because it's so large -- my hope is that there is some way to divvy it up to make it more tractable rather than over-burdening a smaller set of individuals.

File: src/Cello/mesh.CI
Comment (fearmayo) [inline src/Cello/mesh.CI:95]: Should this be here?

**Review by fearmayo — APPROVED** [review]

**Comment by fearmayo** [comment]
> @bwoshea @mabruzzo @fearmayo @jwise77 While we still only need two approvals to merge, I've assigned all four of you to this PR because it's so large -- my hope is that there is some way to divvy it up to make it more tractable rather than over-burdening a smaller set of individuals.

I've had a look/skim through the changes and (without understanding everything) it looks ok. There are a lot of #ifdef BUG_FIX_150 statements which seem like they should be cleaned up? Otherwise though the proof of this PR will be in test runs which as we heard in the last meeting are running. I see no reason not to merge this PR in.

**Comment by jobordner** [comment]
> I've had a look/skim through the changes and (without understanding everything) it looks ok. There are a lot of #ifdef BUG_FIX_150 statements which seem like they should be cleaned up?

Thanks for your review, John. To clarify, the bug fix is a workaround for an apparent memory leak in _Charm++_, not Cello. I wanted to leave the ifdef in until the leak gets fixed in Charm++, and having the compile option to turn it off is an easy way to test it. I will rename "BUG_FIX_150" to something more descriptive and add comments.

File: doc/source/design/design-adapt.rst
Comment (mabruzzo) [inline doc/source/design/design-adapt.rst:134]: The way that ``LevelInfo`` works is pretty self-evident from the class definition. But the way that ``face_level_`` works is much less evident. Would you mind elaborating more on how these encode the same information?

File: doc/source/design/design-adapt.rst
Comment (mabruzzo) [inline doc/source/design/design-adapt.rst:135]: Can you clarify what you mean about the "API for the newer ``LevelInfo`` section"? Aren't you documenting the entire API for the ``Adapt`` class?

File: doc/source/design/design-adapt.rst
Comment (mabruzzo) [inline doc/source/design/design-adapt.rst:140]: Is this only used during initialization? If so, it might be worth mentioning that.

File: src/Cello/io_OutputImage.cpp
Comment (mabruzzo) [reply src/Cello/io_OutputImage.cpp:416]: Can we maybe add a comment explaining why this doesn't need to be initialized?

File: src/Cello/mesh_Hierarchy.hpp
Comment (mabruzzo) [inline src/Cello/mesh_Hierarchy.hpp:254]: Out of curiosity, what is the motivation for transitioning from `bool`s to `int`s? (Are there more than 2 states or is there something else?)

File: src/Cello/simulation_Simulation.cpp
Comment (mabruzzo) [inline src/Cello/simulation_Simulation.cpp:734]: Out of curiosity, were there any problems with the fact that we weren't initializing the periodicity values here in the past? Or did some other change require us to initialize them here?

**Review by mabruzzo — COMMENTED** [review]
Apologies for taking so long to go through this. 

This all generally looks good to me. I got through the documentation and most of the files with minor changes. I would've like to have combed through things in more detail, but there's a lot of code here that I'm generally unfamiliar with

I didn't look at the following 10 files in detail (or at all):

[src/Cello/charm_MsgAdapt.cpp](https://github.com/enzo-project/enzo-e/pull/137/files#diff-395ea1efd19fccb51e57e692a8d5799ded7e3deb57c82f5a674dd37574f5d76d)
[src/Cello/charm_MsgAdapt.hpp](https://github.com/enzo-project/enzo-e/pull/137/files#diff-1d9855fefe92c6da4b043df7a4ce9e7ff353a4ea20b8366c598a4e16f160ba30)
[src/Cello/control_adapt.cpp](https://github.com/enzo-project/enzo-e/pull/137/files#diff-e40f57994d0048d79eb6817780b80091a02c3dc4153859ab0937a3f9ce42aa2a)
[src/Cello/mesh_Adapt.cpp](https://github.com/enzo-project/enzo-e/pull/137/files#diff-6cf20432fc0d20848ff3d023982c5169839707e6aa3bfaf685242839f3bf6b0a)
[src/Cello/mesh_Adapt.hpp](https://github.com/enzo-project/enzo-e/pull/137/files#diff-b7bcf1b1b8e4af0a97d1e67cd25f65f35618663c9886db7bde6bb76aa11d8806)
[src/Cello/mesh_Block.cpp](https://github.com/enzo-project/enzo-e/pull/137/files#diff-8c507d45f4a0b0bf8ec9cd672c51839ba12cdbd82598838db0cb2108809e370a)
[src/Cello/mesh_Block.hpp](https://github.com/enzo-project/enzo-e/pull/137/files#diff-d09f5da714843643a57eb7805522043c27c0ce5e36de6ee93e093f1e2164b25a)
[src/Cello/test_Adapt.cpp](https://github.com/enzo-project/enzo-e/pull/137/files#diff-87e3593745d2536d808b7cea670360a6e9d14ef8451b1c3eb34551f5ebce6323)
[src/Enzo/enzo_EnzoSolverCg.cpp](https://github.com/enzo-project/enzo-e/pull/137/files#diff-bc4baa5f1c6e89a5a10515aaffbb932ef88788543596ff29ca9ef6d59b53e4a9)
[src/Enzo/enzo_EnzoSolverMg0.cpp](https://github.com/enzo-project/enzo-e/pull/137/files#diff-4fd57329af38b152e0029ea28ab2dbc99456e14813abfdeb95d5656219b25b41)

If I have time later this week I'll try to look through them. But if someone else has a chance to go through them, that would be great.

File: src/Cello/io_OutputImage.cpp
Comment (jobordner) [reply src/Cello/io_OutputImage.cpp:416]: This is residual code for debugging; I'll remove it.

File: src/Cello/problem_MethodFluxCorrect.cpp
Comment (jobordner) [reply src/Cello/problem_MethodFluxCorrect.cpp:414]: the index is unnecessary debug code which I'll remove

File: doc/source/design/design-adapt.rst
Comment (jobordner) [reply doc/source/design/design-adapt.rst:9]: Added:

> Blocks are partitioned into "levels" based on how refined they are:
> more highly-refined blocks are in higher-numbered levels, with the
> "root-level" of the simulation defined as "level 0".  The difference
> in resolution between successive levels L and L+1 (the "refinement
> factor") is always two in Enzo-E.

File: doc/source/design/design-adapt.rst
Comment (jobordner) [reply doc/source/design/design-adapt.rst:105]: Revised:

> The balancing step of the algorithm proceeds by alternately sending a
> block's level bounds to its neighbors, and, having received updated bounds from its neighbors,
> updating the block's own level bounds.

File: doc/source/design/design-adapt.rst
Comment (jobordner) [reply doc/source/design/design-adapt.rst:114]: Agreed.

File: doc/source/design/design-adapt.rst
Comment (jobordner) [reply doc/source/design/design-adapt.rst:134]: I don't want to get too technical in the documentation. How about:

> (The "face_level_" representation is a carry-over from the
> previous algorithm, but was retained because it simplifies code that needs to access a
> neighbor's level given the neighbor's relative direction rather than absolute Index).

File: doc/source/design/design-adapt.rst
Comment (jobordner) [reply doc/source/design/design-adapt.rst:135]: I concentrated on the added methods in the API related to the new algorithm, which excludes existing methods related to face_level_. That is, methods below "`//================...`" in `mesh_Adapt.hpp`

File: doc/source/design/design-adapt.rst
Comment (jobordner) [reply doc/source/design/design-adapt.rst:140]: This is only required for testing, since otherwise cello::rank() is used. I'll update hpp comment and docs.

File: src/Cello/main.cpp
Comment (jobordner) [reply src/Cello/main.cpp:89]: I think it's before 7.0.0; I've updated the comments.

File: src/Cello/mesh_Hierarchy.hpp
Comment (jobordner) [reply src/Cello/mesh_Hierarchy.hpp:254]: I _think_ it was because it was previously accessed as int sometimes (e.g. `get_periodicity()` arguments seem to have been `(int *) `), so I changed them all to int for consistency. Though they likely could be all changed to bool.

File: src/Cello/mesh_Index.hpp
Comment (jobordner) [reply src/Cello/mesh_Index.hpp:100]: Updated: " Whether given index is a sibling (has the same parent as this Index)"

File: src/Cello/mesh_Index.hpp
Comment (jobordner) [reply src/Cello/mesh_Index.hpp:107]: "Whether given index is a "nibling" (child of a sibling of this Index)"

File: src/Cello/mesh_Index.hpp
Comment (jobordner) [reply src/Cello/mesh_Index.hpp:169]: Removed "categorize()" since never called.

File: src/Cello/mesh_ItNeighbor.cpp
Comment (jobordner) [reply src/Cello/mesh_ItNeighbor.cpp:84]: deleted

File: src/Cello/simulation_Simulation.cpp
Comment (jobordner) [reply src/Cello/simulation_Simulation.cpp:734]: I don't specifically remember, but according to a log entry I "Replaced `Block::periodicity()` with `Hierarchy::get_periodicity()`" which this looks consistent with.

**Review by mabruzzo — APPROVED** [review]
Thanks for addressing my comments! (sorry for the delay in coming back to this)

Let's merge this in

**Comment by peeples** [comment]
@jobordner It looks like there are just a few conflicts that need to be resolved before this can be merged. Please feel free to merge once these are resolved! Thanks all!

**Pull request merged**
